src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -336,82 +336,73 @@ private SparseFixedBitSet bitSetForSearch()
     @Override
     public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> keys, Expression exp, int limit) throws IOException
     {
-        // VSTODO would it be better to do a binary search to find the boundaries?
-        List<PrimaryKey> keysInRange = keys.stream()
-                                           .dropWhile(k -> k.compareTo(metadata.minKey) < 0)
-                                           .takeWhile(k -> k.compareTo(metadata.maxKey) <= 0)
-                                           .collect(Collectors.toList());
+        // create a sublist of the keys within this segment's bounds
+        int minIndex = Collections.binarySearch(keys, metadata.minKey);
+        minIndex = minIndex < 0 ? -minIndex - 1 : minIndex;
+        int maxIndex = Collections.binarySearch(keys, metadata.maxKey);
+        maxIndex = maxIndex < 0 ? -maxIndex - 1 : maxIndex + 1;
+        List<PrimaryKey> keysInRange = keys.subList(minIndex, maxIndex);
         if (keysInRange.isEmpty())
             return RangeIterator.empty();
-        int numRows = keysInRange.size();
-        logAndTrace("SAI predicates produced {} rows out of limit {}", numRows, limit);
-        if (numRows <= limit)
+
+        logAndTrace("SAI predicates produced {} rows out of limit {}", keysInRange.size(), limit);
+        if (keysInRange.size() <= limit)
             return new ListRangeIterator(metadata.minKey, metadata.maxKey, keysInRange);
 
         int topK = topKFor(limit);
-        var maxBruteForceRows = min(globalBruteForceRows, maxBruteForceRows(topK, numRows));
-        try (PrimaryKeyMap primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap())
+        // if we are brute forcing, we want to build a list of segment row ids, but if not,
+        // we want to build a bitset of ordinals corresponding to the rows.  We won't know which
+        // path to take until we have an accurate key count.
+        SparseFixedBitSet bits = bitSetForSearch();
+        IntArrayList rowIds = new IntArrayList();
+        try (var primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+             var ordinalsView = graph.getOrdinalsView())
         {
-            var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
+            for (PrimaryKey primaryKey : keysInRange)
+            {
+                long sstableRowId = primaryKeyMap.exactRowIdForPrimaryKey(primaryKey);
+                // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
```

Comment:
I thought the PKM was for the whole sstable, and could therefore return row ids for rows that do not have an indexed vector.